### ISSUE TITLE
RabbitMQ boot failure related to feature flags

### ISSUE THREAD

**Issue by Rmarian** [rabbitmq/rabbitmq-server#4881]
Hi RabbitMQ contributors,

We have discovered recently an issue in RabbitMQ.

**Rabbitmq version used**: 3.8.27, Erlang 24.2
**Setup**: three node cluster (each node inside a docker container) with partition handling set to pause_minority 
**Summary of problem**: after shutting down two of the nodes, then starting up one of them and that node fails to write its feature flags to disk then the third node sometimes fails to boot and remains in a zombi state 

**Steps to reproduce:**
1. Ensure following feature flags are enbaled: virtual_host_metadata, user_limits and quorum_queue
1. Stop node 1 and 2.  At this time node 3 should be stopped automatically due to pause_minority condition
2. On node 1 remove feature flag file under /usr/lib/rabbitmq/var/lib/rabbitmq/mnesia
3. Start up node 1 only

**Expected result**:

1. Node 3 starts-up without error

**Actual result**:

1. Sometimes node three fails to boot with the errorr "Error during startup: {error,{incompatible_feature_flags,{not_active,"All replicas on diskfull nodes are not active yet"..."
2. After this error is logged the rabbitmq process hangs with no exit and no crash report written to disk

Attached full error log for node 3: [rabbitmq-error.txt](https://github.com/rabbitmq/rabbitmq-server/files/8751900/rabbitmq-error.txt)

**Additional nodes/remarks**:
1. The issue does not happen all the time, it is sort of a race condition which I explain in point 3 below.
2. The reason why the feature flags sometimes fails to be written for node 1 or 2 I think is because we are doing failover tests that involve hard reset/power-off of the hosts where rabbitmq is running
3. My assumption on the sequence of events that leads to this is that, node 1 upon startup detects that there are no feature flags written to disk (I see the warning "the previous instance of this node must have failed to write the `feature_flags` file at `/usr/lib/rabbitmq/var/lib/rabbitmq/mnesia"), then it defaults for a short time to no flags enabled until it restores the previous set flags. If during this time window the the third node that is coming up sees that node 1 has no flags enabled and tries to activate for example flag "user_limits" we get into this error because the activation of this feature flag requires all nodes to be running (node 2 is not running at this point)
4. I would expect at the very least that rabbitmq node three should exit gracefully and write an appropriate crash report in this situation so that we have the opportunity of automatically detecting this issue and restarting the process

Thanks,
Radu.

**Comment by michaelklishin** [comment]
RabbitMQ 3.8 is [out of support](https://www.rabbitmq.com/versions.html) and now only gets security-related fixes and a limited number of dependency updates.

If you can reproduce with 3.10.2, let us know.

**Comment by michaelklishin** [comment]
Relevant lines from the log:

```
2022-05-23T05:44:52.326386Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; BOOT FAILED;
2022-05-23T05:44:52.326403Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; ===========;
2022-05-23T05:44:52.326419Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; Error during startup: {error,;
2022-05-23T05:44:52.326436Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/];                           {incompatible_feature_flags,;
2022-05-23T05:44:52.326451Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/];                               {not_active,;
2022-05-23T05:44:52.326466Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/];                                   "All replicas on diskfull nodes are not active yet",;
2022-05-23T05:44:52.32648Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/];                                   rabbit_queue,;
2022-05-23T05:44:52.326495Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/];                                   ['rabbit@rabbitmq-2.xvp.frequentis.frq']}}};
2022-05-23T05:44:52.326512Z;ERR;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; ;
2022-05-23T05:44:52.336031Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.324 [error] <0.3214.0>                           {incompatible_feature_flags,#015;
2022-05-23T05:44:52.336077Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.324 [error] <0.3214.0>                               {not_active,#015;
2022-05-23T05:44:52.336091Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.324 [error] <0.3214.0>                                   "All replicas on diskfull nodes are not active yet",#015;
2022-05-23T05:44:52.336102Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.325 [error] <0.3214.0>                                   rabbit_queue,#015;
2022-05-23T05:44:52.33611Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.325 [error] <0.3214.0>                                   ['rabbit@rabbitmq-2.xvp.frequentis.frq']}}}#015;
2022-05-23T05:44:52.336119Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:52.325 [error] <0.3214.0> #015;
2022-05-23T05:44:53.333691Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:53.326 [error] <0.3213.0> CRASH REPORT Process <0.3213.0> with 0 neighbours exited with reason: {{incompatible_feature_flags,{not_active,"All replicas on diskfull nodes are not active yet",rabbit_queue,['rabbit@rabbitmq-2.xvp.frequentis.frq']}},{rabbit,start,[normal,[]]}} in application_master:init/4 line 142#015;
2022-05-23T05:44:53.333728Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:53.327 [info] <0.3213.0> [{initial_call,{application_master,init,['Argument__1','Argument__2','Argument__3','Argument__4']}},{pid,<0.3213.0>},{registered_name,[]},{error_info,{exit,{{incompatible_feature_flags,{not_active,"All replicas on diskfull nodes are not active yet",rabbit_queue,['rabbit@rabbitmq-2.xvp.frequentis.frq']}},{rabbit,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,142}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}},{ancestors,[<0.3212.0>]},{message_queue_len,1},{messages,[{'EXIT',<0.3214.0>,normal}]},{links,[<0.3212.0>,<0.44.0>]},{dictionary,[]},{trap_exit,true},{status,running},{heap_size,1598},{stack_size,28},{reductions,464}], []#015;
2022-05-23T05:44:53.333739Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:53.327 [info] <0.44.0> Application rabbit exited with reason: {{incompatible_feature_flags,{not_active,"All replicas on diskfull nodes are not active yet",rabbit_queue,['rabbit@rabbitmq-2.xvp.frequentis.frq']}},{rabbit,start,[normal,[]]}}#015;
2022-05-23T05:44:53.333826Z;INFO;dockerhost2.xvp.frequentis.frq;P3483/;[rabbitmq-3/]; 2022-05-23 05:44:53.327 [info] <0.44.0> Application rabbit exited with reason: {{incompatible_feature_flags,{not_active,"All replicas on diskfull nodes are not active yet",rabbit_queue,['rabbit@rabbitmq-2.xvp.frequentis.frq']}},{rabbit,start,[normal,[]]}}#015;
```

**Comment by michaelklishin** [comment]
I chatted with the team and it was immediately pointed out that you remove the feature flag state file on one of the nodes. Deleting data files produced by nodes will cause issues, inconsistent behavior, and should not be done in 99.9% of environments.

**Comment by dumbbell** [comment]
@Rmarian: Hi! I see in the steps to reproduce that you explicitly removes the `feature_flags` file. Did you encounter a case where the file was actually removed?

This file is a critial part of the RabbitMQ state and if it's missing, then there is little chance that the node will work properly. In particular in a cluster, feature flags compatibility is verified before the Mnesia database is started because a feature flag could change the database schema. That explains why node 1 and 3 fail to form a cluster again.

Also, when reporting an issue with feature flags, could you please attach the logs of both nodes? There should be relevant messages from the feature flags subsystem on both sides.

**Comment by Rmarian** [comment]
Hi @dumbbell ,

Yes, on one of the nodes I always see the following warning when the crash happens:  [warning] <0.280.0> Feature flags: the previous instance of this node must have failed to write the `feature_flags` file at `/usr/lib/rabbitmq/var/lib/rabbitmq/mnesia/...`"

I managed to reproduce this warning by manually deleting the flag file under /usr/lib/rabbitmq/var/lib/rabbitmq.

We have requirements related to failover and to test these scenarios we do a lot of hard power-offs/resets of the hosts so I believe that from time to time this file gets corrupted because of this and that triggers this warning

Even with this file missing/corrupted, the node somehow knows to restore the old enabled feature flags. But until it does so there are 15 seconds where it starts up with node flags enabled by default. You can see this in the log snippet attached below

Detailed startup logs for the second node: [rabbitmq-error-missing-feature-flags.txt](https://github.com/rabbitmq/rabbitmq-server/files/8753027/rabbitmq-error-missing-feature-flags.txt).

**Comment by Rmarian** [comment]
Hi @michaelklishin ,

Actually what we observe in production is the following warning right before the issue happens:

**Comment by Rmarian** [comment]
Hi @michaelklishin , @dumbbell ,

An update on this topic: I found out why the feature flag file was being deleted:

We are running rabbitmq inside docker container and we configured a persistent volume for the rabbitmq data however the feature flag file was being stored outside the volume in the container internal storage. Every time we would restart the host the container would be re-created and the feature flag would of course be lost.

After I moved the file to the volume the problem seems to be resolved now.

Thanks,
Radu.
